[step2] 블랙잭

pkch93 opened the pull request:
안녕하세요 :)

2단계 블랙잭 미션 구현해서 PR 올립니다.

참고로 제가 11.28 ~ 12.02까지 여행 일정이 있어서,, 리뷰해주시면 그 이후에나 볼 수 있을거 같습니다.

천천히 봐주시면 될거 같아요 :) 잘부탁드립니다.

pkch93 committed:
[step1] hand-on DSL 실습

pkch93 committed:
[step1] 스킬 정보 DSL 추가

pkch93 committed:
[step1] 언어 정보 DSL 추가

pkch93 committed:
[step2] 블랙잭 구현 필요 사항 README.md에 정리

pkch93 committed:
[step2] 블랙잭 구성요소 기본 디자인

pkch93 committed:
[step2] CardTest 추가

pkch93 committed:
[step2] CardTest 통과하도록 구현

pkch93 committed:
[step2] HandsTest 추가

pkch93 committed:
[step2] HandsTest 통과하도록 구현

pkch93 committed:
[step2] Card 클래스 정의

pkch93 committed:
[step2] Player, Hands 클래스 외부로 분리

pkch93 committed:
[step2] Hand 제거 및 Player#addCard 테스트 및 구현

pkch93 committed:
[step2] 초기 2장 카드 뽑기 기능 추가

pkch93 committed:
[step2] 카드를 뽑을 플레이어 조회 테스트 추가

pkch93 committed:
[step2] 카드 추가 기능 구현

pkch93 committed:
[step2] 블랙잭 게임 결과 추출 테스트 추가

pkch93 committed:
[step2] 블랙잭 게임 결과 추출 구현

pkch93 committed:
[step2] 결과 출력 구현

pkch93 committed:
[step2] 출력문구 정리

pkch93 committed:
[step2] 플레이어가 카드를 뽑을때 카드의 총 합계가 21 이상이면 카드 뽑지 않는 테스트 케이스 추가

pkch93 committed:
[step2] 플레이어가 카드를 뽑을때 카드의 총 합계가 21 이상이면 카드 뽑지 않는 기능 구현

wisemuji commented on README.md:
기능 목록을 잘 정리해주셨습니다 👍 
이상적으로는 이렇게 작성하신 기능 목록이 결국에는 커밋 단위, 테스트 시나리오까지 이어지면 좋습니다.

wisemuji commented on src/main/kotlin/BlackJackGame.kt:
테스트 환경에서 참조하기 편하도록 `package blackjack`과 같이 블랙잭 프로그램을 패키지로 감싸주세요!

pkch93 replied:
fixed 028c02ae

wisemuji commented on src/main/kotlin/BlackJackGame.kt:
```diff
@@ -0,0 +1,67 @@
+    fun initialDraw(): List<DrawResult> {
+        (1..2).forEach { _ ->
+            players.forEach { it.addCard(deck.draw()) }
+        }
```
Kotlin의 [repeat](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/repeat.html)을 활용하시면 어떨까요~?

pkch93 replied:
fixed 6f32c92a

wisemuji commented on src/main/kotlin/BlackJackGame.kt:
```diff
@@ -0,0 +1,67 @@
+    val playerName: String,
+    val cards: List<DrawCard>,
+    val totalValue: Int,
+)
```
https://kotlinlang.org/docs/coding-conventions.html#source-file-organization
> Placing multiple declarations (classes, top-level functions or properties) in the same Kotlin source file is encouraged as long as these declarations are closely related to each other semantically, and the file size remains reasonable (not exceeding a few hundred lines).

공식 문서에 따르면, 동일한 파일에 여러 개의 클래스를 배치하는 것은 의미적으로 밀접하게 관련되어 있는 경우 권장됩니다. `DrawResult`와 `BlackJackGameResult`가 의미적으로 밀접하게 관련이 없다면, 이를 별도의 파일로 분리하면 어떨까요?

pkch93 replied:
"의미적으로 밀접하게 관련"이라는게 사람마다 해석하기 나름일 것 같은데요.

일단 제가 `BlackJackGame`에 `DrawResult`와 `BlackJackGameResult`를 둔 이유는 두 Result 클래스 모두 `BlackJackGame`의 동작 결과물이고 결과물을 한 파일내에서 바로 볼 수 있는 것이 유의미하다고 생각했기 때문입니다. 물론 `BlackJackGame`이 비대하여 `DrawResult`와 `BlackJackGameResult`를 파악하기 힘들다면 분리해도 무방할 것 같지만 굳이 그렇게 하지는 않았습니다.

리뷰어님께서는 아마 "의미적으로 밀접하지 않다"라고 생각하신거 같아요. 위와 같이 의견을 주신 이유가 궁금합니다.

wisemuji replied:
말씀하신 것처럼 "의미적으로 밀접하게 관련"이라는게 사람마다 해석하기 나름일 것 같습니다.
제가 의견드렸던 이유는 `BlackJackGame`에서만 `DrawResult`와 `BlackJackGameResult`을 참조하지 않기 때문입니다.

말씀하신 것처럼 두 클래스를 `BlackJackGame`의 동작 결과물로 볼 수는 있지만, 이 결과물들이 다른 파일에서도 자주 참조되며 활용될 가능성이 있다면, 해당 클래스들을 별도의 파일로 분리하여 다른 개발자들도 쉽게 찾을 수 있도록 하는 것이 적절하다고 생각했습니다.

저의 주관적인 의견이니 경철님의 말씀도 충분히 합리적이라고 생각합니다.
이 코멘트 반영 여부는 경철님이 판단해주셔도 좋을 것 같아요 🙂

wisemuji commented on src/main/kotlin/BlackJackGame.kt:
```diff
@@ -0,0 +1,67 @@
+                cards = it.currentCards,
+                totalValue = it.totalValue(),
+            )
+        }
```
BlackJackGame에서 View의 참조를 제거하여 controller가 비대해지지 않을 수 있게 되었어요! 👍 
다만 현재는 View에서 직접 BlackJackGame의 함수를 호출하며 게임을 제어하고 있어서 아래 그림과는 일치하지 않는 구조일 것 같아요.

![image](https://github.com/user-attachments/assets/c58b19c3-f78d-4136-b552-1140cad02b1b)


또한 앞으로 요구 사항이 많아질수록 BlackJackGame 역시 커질 수 있을 것 같은데요,
적절한 객체의 책임으로 위임할 수 있는 로직이 있을까요?

pkch93 replied:
GameController라는 중간 매개체를 두어 분리시켰습니다.

fixed 08dbc754

wisemuji commented on src/main/kotlin/Player.kt:
```diff
@@ -0,0 +1,35 @@
+        cards.add(card)
+
+        if (totalValue() >= 21) {
+            stopDraw()
```
다음을 참고하여 블랙잭 게임에 대한 이해도를 높여보면 어떨까요?
- https://www.high1.com/casino/contents.do?key=1764
- https://namu.wiki/w/블랙잭(카드게임)

예를 들어 bust라는 도메인 용어를 적용해볼 수 있을 것 같아요 🙂

pkch93 replied:
fixed 17c2c989

wisemuji commented on src/main/kotlin/Card.kt:
```diff
@@ -0,0 +1,74 @@
+            CardNumber.JACK -> "J"
+            CardNumber.QUEEN -> "Q"
+            CardNumber.KING -> "K"
+        }
```
도메인 객체에 뷰 요구사항이 포함되는게 적절할까요? 자동차 경주 5단계 요구사항을 참고하여 도메인 로직과 뷰 로직을 명확하게 분리하고, 의존 방향을 정리해보시면 좋을 것 같아요 🙂 

![image](https://github.com/user-attachments/assets/ccf970ed-3ebd-4405-b95a-ab2e6fb76e3e)

pkch93 replied:
카드 표현을 위한 View용 객체 `CardExpression`을 만들어서 분리했어요.

fixed 64e66d2f

wisemuji commented on src/main/kotlin/Card.kt:
```diff
@@ -0,0 +1,74 @@
+    KING({ 10 });
+
+    fun value(calculatedValue: Int): Int = decideCardValue(calculatedValue)
+}
```
IntelliJ의 다음 설정 옵션을 활성화하여 EOF 경고를 없애주세요!

```
Settings → Editor → General → Ensure every saved file ends with line break
```

wisemuji commented on src/main/kotlin/Deck.kt:
```diff
@@ -0,0 +1,9 @@
+        .shuffled()
+        .toMutableList()
+
+    fun draw(): Card = cards.removeAt(0)
```
Deck이 실제로 카드를 잘 Draw하고 있는지 검증하는 테스트를 작성해보면 어떨까요?
다만 현재는 `shuffled()`라는 랜덤 로직을 테스트 과정에서 참조하게 될텐데요, 이를 개선하는 작업이 필요할 것 같아요.

다음 글이 도움이 되었으면 좋겠습니다.
- https://tecoble.techcourse.co.kr/post/2021-10-04-strategy-command-pattern/
- https://tecoble.techcourse.co.kr/post/2020-05-07-appropriate_method_for_test_by_parameter/

pkch93 replied:
fixed 63780638

wisemuji commented on src/main/kotlin/Player.kt:
```diff
@@ -0,0 +1,35 @@
+    val currentCards: List<DrawCard>
+        get() = cards
+            .map { it.toDrawCard() }
+            .toList()
```
toList()를 활용한 방어적 복사 👍  하지만 의도와는 다르게 완전히 변경사항 전파를 끊어내지 못합니다. 다음과 같이 작성하면 외부에서 리스트를 조작할 수 있게 됩니다.

```kotlin
    val cards = mutableListOf<Card>(...)
    val player = Player(cards = cards, ...)
    println(player.currentCards.size) // 1
    cards.removeAt(0)
    println(player.currentCards.size) // 0
```

따라서 생성자로 Mutable 타입을 받을 때에는 항상 주의해야 합니다. 어떻게 개선할 수 있을까요?

pkch93 replied:
참조를 바꿔줘야겠군요.

fixed 297a7660

wisemuji commented on src/test/kotlin/PlayerTest.kt:
```diff
@@ -0,0 +1,101 @@
+                Player(
+                    PlayerName("테스트"),
+                    mutableListOf(
+                        Card(
+                            CardSuit.HEARTS,
+                            CardNumber.TWO,
+                        ),
+                        Card(
+                            CardSuit.SPADES,
+                            CardNumber.THREE,
+                        ),
+                        Card(
+                            CardSuit.DIAMONDS,
+                            CardNumber.FOUR,
+                        ),
+                        Card(
+                            CardSuit.CLUBS,
+                            CardNumber.ACE,
+                        ),
```
객체 단위 별 의미 있는 테스트 시나리오를 잘 만들어주셨네요 👍 

장기적으로 프로그램이 커질수록 픽스쳐로 관리하고 싶은 샘플 코드들이 생길텐데요, 이러한 픽스쳐는 어떻게 관리하면 좋을지 스스로 고민해봐도 좋아요!
고민에 도움이 되도록 관련 예시를 첨부합니다.
- https://thalals.tistory.com/375
- https://github.com/woowacourse/service-apply/blob/master/src/test/kotlin/apply/JudgmentFixtures.kt

wisemuji replied:
더불어 같이 고민해보시면 좋을 내용은, "과연 블랙잭 게임에서 생성될 수 있는 카드의 경우의 수가 몇 개인지" 입니다. 
매번 Card 인스턴스를 새로 만들어야 할까요?
(꼭 테스트 코드에서만 해당되는 내용이 아닐 것 같네요!)

pkch93 replied:
픽스쳐를 만들까도 고민했었는데 굳이 불필요하다고 생각했었어요. 

픽스쳐로 뽑아냈을때 표현이 지금이랑 크게 다르지 않다고 생각이 들었고, 지금 형태도 충분히 Card가 어떤게 담겼는지 잘 표현해주고 있다고 생각하구요.

지금같은 형태에서는 `aceClubsCard()` 이런 식으로 픽스쳐가 나올거 같은데 `Card(CardSuit.CLUBS, CardNumber.ACE)`와
 크게 다르지 않은 형태라고 생각합니다. 테스트마다 사용되는 공통점이 보이는 경우에 의미를 담은 픽스처를 따면 좋을거 같네요.

pkch93 replied:
+ 매번 인스턴스를 새로 만들 필요는 없지만 특히 테스트에서는 굳이 새로 인스턴스를 만들지 말아야하는 것에 대해 고민해야할 필요가 있나 싶네요.

wisemuji replied:
> 매번 인스턴스를 새로 만들 필요는 없지만 특히 테스트에서는 굳이 새로 인스턴스를 만들지 말아야하는 것에 대해 고민해야할 필요가 있나 싶네요.

그래서 꼭 테스트 코드에서만 해당되는 내용이 아닐 것 같다고 말씀드린 거였습니다!

> 플레잉 카드(= 트럼프)는 일반적으로 52장이며, 모양별로 4종류의 슈트(다이아몬드(◆), 하트(♥), 클로버(♣), 스페이드(♠))와 13개의 끗수(denomination)로 이루어져 있다.

게임에서 사용되는 카드의 종류가 제한되어 있다면, 굳이 매번 생성하지 않더라도, 미리 생성해두고 해당 인스턴스를 참조하도록 개선해볼 수 있을 것 같아요. 이 구조가 무조건 장점만 있지는 않을 수 있습니다. 단, 이 과정 모두가 학습을 위한 미션인 만큼, 최대한 리뷰이분이 여러 방법을 시도해보시고 각각의 장단점을 비교하셨으면 좋겠어서 드린 피드백입니다.

강사님께서 로또 피드백 수업에서도 해당 주제의 내용을 다루신 것으로 알고 있습니다.
아마 다음주 화요일 라이브 수업(블랙잭 피드백)에서도 다뤄질 것 같은데요, 이 수업도 더욱 도움이 되셨으면 좋겠습니다!

wisemuji reviewed (changes requested):
안녕하세요 경철님!

2단계 구현을 너무 잘 해주셨습니다 💯  깔끔한 구현이 좋았습니다
다음 단계로 넘어가기 전에 먼저 고민해보시면 좋을만한 코멘트를 달았으니 확인해주시면 감사하겠습니다.

천천히 리뷰해도 괜찮다고 배려해주셔서 감사드립니다 🙇 
(하지만 진짜 천천히 리뷰하면 재미 없으니 그냥 빨리 리뷰 했습니다 ㅋㅋㅋ)

pkch93 committed:
[step2] blackjack 패키지로 관련 클래스 이동

pkch93 committed:
[step2] 코드리뷰 반영

- repeat 활용
- https://github.com/next-step/kotlin-blackjack/pull/725#discussion_r1861709933

pkch93 committed:
[step2] Player의 cards 프로퍼티를 새로운 List를 생성하여 사용하도록 수정

pkch93 committed:
[step2] 뷰에서 카드를 표현하는 객체 CardExpression 정의

- https://github.com/next-step/kotlin-blackjack/pull/725#discussion_r1861727904

pkch93 committed:
[step2] Deck 생성 형태 수정

- https://github.com/next-step/kotlin-blackjack/pull/725#discussion_r1861731691

pkch93 committed:
[step2] 총점 21점이 넘는 경우를 bust라는 도메인 용어 사용하도록 수정

pkch93 committed:
[step2] domain 패키지 추가

pkch93 committed:
[step2] GameController 추가하고 InputView에서 도메인 로직 분리

pkch93 committed:
[step2] 누락된 초기 카드 뽑기 결과 출력 추가

wisemuji commented on README.md:
게임에 몇 가지 버그가 있는 것 같아요!

<img width="441" alt="image" src="https://github.com/user-attachments/assets/e166d7a3-0aac-47d6-a06f-02a31163fa03">

- 세 번째 플레이어의 카드 드로잉 중 첫 번째 플레이어의 카드 드로잉이 발생함
- A가 1로 계산되어야 유리한 상황에서 11로 계산됨

wisemuji commented on src/main/kotlin/blackjack/domain/Player.kt:
```diff
@@ -0,0 +1,40 @@
+    private var draw: Boolean = true,
+    cards: List<Card> = listOf(),
+) {
+    private val cards: MutableList<Card> = cards.toMutableList()
```
방어적 복사 👍

wisemuji commented on src/main/kotlin/blackjack/view/CardExpression.kt:
```diff
@@ -0,0 +1,43 @@
+        fun of(
+            drawCard: DrawCard,
+        ): CardExpression =
+             CardExpression("${convertNumber(drawCard.number)}${convertSuit(drawCard.suit)}")
```
코드 리뷰를 잘 반영해주셨습니다 👍

wisemuji reviewed (approved):
안녕하세요 경철님! 리뷰 코멘트가 많았는데 반영 잘 해주셨습니다!

코멘트별 커밋을 남겨주셔서 덕분에 리뷰하기에 편했습니다 🙇 
남겨주신 의견에 대한 답변도 남겼으니 경철님만의 답을 내리셨으면 좋겠습니다.

다음 미션도 화이팅입니다!
